segmentstore/storage/impl/src/main/java/io/pravega/segmentstore/storage/impl/bookkeeperstorage/StorageLogManager.java

```diff
@@ -0,0 +1,813 @@
+            return retVal;
+        }
+
+        lh.asyncAddEntry(bytes, (errorCode, ledgerHandle, l, future) -> {
+            if (errorCode == 0) {
+                retVal.complete(null);
+            } else {
+                log.warn("asyncAddEntry failed for ledger with ledger id {}, LAC {}, LAP {} ", lh.getId(), lh.getLastAddConfirmed(), lh.getLastAddPushed());
+                retVal.completeExceptionally(translateBKException(BKException.create(errorCode), segmentName));
+            }
+        }, retVal);
+
+        return retVal;
+    }
+
+    private CompletableFuture<List<CuratorOp>> getZKOperationsForConcat(String segmentName, String sourceSegment, long offset) {
+        return this.getOrRetrieveStorageLedger(segmentName, false)
+                   .exceptionally(exc -> {
+                               translateZKException(segmentName, exc);
+                               return null;
+                           }
+                   )
+                   .thenCombine(this.getOrRetrieveStorageLedger(sourceSegment, false),
+                           (target, source) -> {
+                               if (!source.isSealed()) {
+                                   throw new IllegalStateException("source must be sealed");
+                               }
+                               if (source.getContainerEpoch() != this.containerEpoch) {
+                                   throw new CompletionException(new StorageNotPrimaryException(target.getName()));
+                               }
+                               List<CuratorOp> operations = new ArrayList<>();
+                               LedgerData lastLedger = target.getLastLedgerData();
+                               if (lastLedger.getLedgerHandle().getLength() == 0) {
+                                   operations.add(createLedgerDeleteOp(lastLedger, target));
+                               }
+                               List<CuratorOp> retVal = target.addLedgerDataFrom(source);
+                               operations.addAll(retVal);
+                               // Update the segment also to ensure fencing has not happened
+                               operations.add(createLedgerUpdateOp(target));
+                               return operations;
+                           });
+    }
+
+    private CuratorOp createLedgerDeleteOp(LedgerData lastLedger, StorageLog target) {
+        return zkClient.transactionOp().delete()
+                       .forPath(getZkPath(target.getName()) + "/" + lastLedger.getStartOffset());
+    }
+
+    private CuratorOp createLedgerUpdateOp(StorageLog target) {
+        return zkClient.transactionOp().setData().withVersion(target.getUpdateVersion())
+                       .forPath(getZkPath(target.getName()), target.serialize());
+    }
+
+    //endregion
+
+    //region ZK private helper methods
+
+    private void translateZKException(String streamSegmentName, Throwable exc) {
```

Comment:
To config, are you sure these async calls throw an exception? The ZK async API has return codes, it doesn't throw exceptions. In the case we keep the async wrapper, please make sure we are using it properly.

Reply:
Yes, I have verified that these exceptions are actually thrown.